### PR TITLE
Fix capturing of a generic type in local functions

### PR DESCRIPTION
**Pull request by agocke**
This changes our generic-lowering logic to use 'Arity' instead of
'IsGenericType' to figure out whether or not a local function frame
needs generic subsitution. We previously used IsGenericType, which is
inappropriate because it also considers whether or not the containing
type of the frame is generic. The containing type is necessarily already
subsituted when calling a local function, so it will never be possible
to substite type parameters.

Instead, we should use Arity of the frame, which is only non-zero if the
frame itself is generic, regardless of the containing type.

### CODE REVIEW

**Comment by agocke** [comment]
\+ @dotnet/roslyn-infrastructure It looks like some random tests are failing with AccessViolations. Has anything like this been reported before?

**Comment by agocke** [comment]
test this please

**Comment by agocke** [comment]
test vsi please

**Comment by agocke** [comment]
test mac please

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs [hunk 1]
```diff
@@ -30,6 +30,114 @@ public static IMethodSymbol FindLocalFunction(this CommonTestBase.CompilationVer
     public class CodeGenLocalFunctionTests : CSharpTestBase
     {
         [Fact]
+        public void VS15Pre3BugTest()
```
Comment (VSadov) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs]: test name could be a bit more informative

Comment (agocke) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs]: Yeah, that was my investigation name, will change :)

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs [hunk 2]
```diff
@@ -30,6 +30,114 @@ public static IMethodSymbol FindLocalFunction(this CommonTestBase.CompilationVer
...
+}
+class C2
+{
+    public static void Main(string[] args) => new C<int>().M(10);
+}
+";
+            VerifyOutput(src, @"0
+10
+0");
+        }
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs]: Should nested generic functions be tested as well?

**Comment by cston** [comment]
LGTM

File: src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs (hunk 2 above)
Comment (agocke) [inline src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenLocalFunctionTests.cs]: I see at least one test, `StructClosureGeneric`, in the same file that seems to test that. Was there something specific about the nesting that I should test?

**Comment by VSadov** [comment]
LGTM

**Pull request merged**